### PR TITLE
Init `RuntimeLogger` automatically for each runtime api call

### PR DESCRIPTION
**Pull request by bkchr**
This pr change the runtime api in such a way to always and automatically
enable the `RuntimeLogger`. This enables the user to use `log` or
`tracing` from inside the runtime to create log messages. As logging
introduces some extra code and especially increases the size of the wasm
blob. It is advised to disable all logging completely with
`sp-api/disable-logging` when doing the wasm builds for the on-chain
wasm runtime.

Besides these changes, the pr also brings most of the logging found in
frame to the same format "runtime::*".

Fixes: https://github.com/paritytech/substrate/issues/7927

Should fix: https://github.com/paritytech/substrate/issues/8107

polkadot companion: https://github.com/paritytech/polkadot/pull/2522

### CODE REVIEW

**Comment by kianenigma** [comment]
Maybe we can do something like [this](https://github.com/paritytech/substrate/pull/8123/files#diff-6faadaafc4a2f7556e8ba9cf9bd8544e5c039e97441e1d470a88c296d39819edR1378-R1382) to also standardize this `runtime::` prefix.

File: frame/staking/src/lib.rs [hunk 1]
```diff
@@ -3189,29 +3189,29 @@ impl<T: Config> Module<T> {
 /// some session can lag in between the newest session planned and the latest session started.
 impl<T: Config> pallet_session::SessionManager<T::AccountId> for Module<T> {
 	fn new_session(new_index: SessionIndex) -> Option<Vec<T::AccountId>> {
-		frame_support::debug::native::trace!(
-			target: LOG_TARGET,
-			"[{}] planning new_session({})",
+		log!(
+			trace,
```
[the remark this answers is not in the document]
Comment (kianenigma) [reply frame/staking/src/lib.rs:3193]: the `log!` is a local macro that reads the target from a crate-local const. I actually recommended above doing the same for all PRs :D we can generate the const via the macros, and enforce the `runtime::` prefix in them as well.

[the remark this answers is not in the document]
Comment (thiolliere) [reply frame/staking/src/lib.rs:3193]: ah ok

File: test-utils/runtime/Cargo.toml [hunk 2]
```diff
@@ -97,3 +97,5 @@ std = [
 	"sp-transaction-pool/std",
 	"trie-db/std",
 ]
+# Special feature to disable logging
+disable-logging = [ "sp-api/disable-logging", "log/max_level_error" ]
```
Comment (thiolliere) [inline test-utils/runtime/Cargo.toml:101]: why is it "log/max_level_error" here instead of "log/max_level_off" don't we want to disable logging ?

**Review by thiolliere — COMMENTED** [review]
If I understand correctly the feature "max_level_off" propagate to all crate thus the wasm binary compiled with feature "disable-logging" compiles all the crate with "log/max_level_off", isn't it?

I'm not familiar with how log removes the logging code.
Or maybe what you mean by:

> As logging introduces some extra code and especially increases the size of the wasm
blob.

Is just about the initialization of logger in wasm not the logging in pallets themself.


Anyway I think it should better to have the feature disable-logging in node-runtime as well so people can compile wasm binary with `cargo build --feature disable-logging -p node-runtime` (or I don't know what it the command documented to build the wasm binary)

File: test-utils/runtime/Cargo.toml (hunk 2 above)
Comment (bkchr) [reply test-utils/runtime/Cargo.toml:101]: That was me trying 3 hours to get this running, until I realized why it doesn't work.. Actually we need to remove this. Good find! :) 
```suggestion
disable-logging = [ "sp-api/disable-logging" ]
```

**Comment by bkchr** [comment]
> If I understand correctly the feature "max_level_off" propagate to all crate thus the wasm binary compiled with feature "disable-logging" compiles all the crate with "log/max_level_off", isn't it?

Yes that is correct.

> I'm not familiar with how log removes the logging code.

The logging is just a macro that is essentially wrapped in some `if`: https://github.com/rust-lang/log/blob/master/src/macros.rs#L201

This feature sets `STATIC_MAX_LEVEL` to `OFF`. So, the if condition is always `false` and the compiler should remove the if completely, because it would be never called. :)

**Comment by bkchr** [comment]
> Another note: I think it should better to have the feature disable-logging in node-runtime as well so people can compile wasm binary with `cargo build --feature disable-logging -p node-runtime` (or I don't know what it the command documented to build the wasm binary)

Yeah in the end that will come. However, I'm not sure if we require this for the Substrate node. In the end I want a proper tutorial that explains this to people. :)

File: test-utils/runtime/build.rs [hunk 3]
```diff
@@ -26,5 +26,13 @@ fn main() {
...
-		.build()
+		.build();
+
+	WasmBuilder::new()
+		.with_current_project()
+		.export_heap_base()
+		.import_memory()
+		.set_file_name("wasm_binary_logging_disabled.rs")
+		.enable_feature("disable-logging")
+		.build();
```
Comment (thiolliere) [inline test-utils/runtime/build.rs:37]: FIY I tried to debug the test `ensure_runtime_logger_works`, I removed these lines (while I already had the file `wasm_binary_logging_disabled.rs` generated so compilation could work).
And the test finally worked.

I think there is some conflict here somehow.

Comment (bkchr) [reply test-utils/runtime/build.rs:37]: Ahh I did not see that this test failed. I will fix it. :)

**Comment by thiolliere** [comment]
So this will need the polkadot companion for adding features to the runtimes,
And also document the process somehow,

But otherwise looks good to me

**Comment by bkchr** [comment]
Yeah. I want to go some 2 way documentation way. For polkadot I will communicate this internal. For Substrate in general I want to have this on the dev hub

**Review by thiolliere — APPROVED** [review]
ok, don't know in which order we do stuff for documentation, but this PR is ok

File: frame/support/Cargo.toml [hunk 4]
```diff
@@ -56,8 +54,7 @@ std = [
 	"sp-staking/std",
 	"sp-state-machine",
 	"frame-support-procedural/std",
+	"log/std",
 ]
-nightly = []
```
Comment (kianenigma) [inline frame/support/Cargo.toml:60]: ahh yes thank you

**Review by kianenigma — APPROVED** [review]
1. I couldn't find this in the wasm-builder: can we have a test to compare the wasm size when the feature is on or off? 

2. As recommended elsewhere, I think can autogenerate the `const LOG_TARGET: ...` for all pallets in the macro, instead of assuming that they all prefix with runtime. We can also generate one helper macro for logging with `LOG_TARGET` (what I have done in staking). Could be an easy follow up.

**Comment by kianenigma** [comment]
And, if we had changelogs, this def. needs to go into one as I think it breaks many downstream projects.

**Comment by bkchr** [comment]
> * I couldn't find this in the wasm-builder: can we have a test to compare the wasm size when the feature is on or off?
> 

For Kusama the difference is 700 bytes :D So, not that much xD 

> * As recommended elsewhere, I think can autogenerate the `const LOG_TARGET: ...` for all pallets in the macro, instead of assuming that they all prefix with runtime. We can also generate one helper macro for logging with `LOG_TARGET` (what I have done in staking). Could be an easy follow up.

Yeah, my plan ;)

**Comment by athei** [comment]
Are the released runtimes compiled with `disable-logging` enabled? I am asking because this feature does not seem to be the default and in the release notes is no command line which was used to build the runtime.

**Comment by bkchr** [comment]
They are disabled for Polkadot/Kusama

**Pull request merged**